lib/grp.gd

```diff
@@ -629,8 +629,13 @@ InstallTrueMethod( IsPerfectGroup, IsGroup and IsTrivial );
 ##
 DeclareProperty( "IsSimpleGroup", IsGroup );
 InstallTrueMethod( IsGroup and IsNonTrivial, IsSimpleGroup );
+
 DeclareSynonymAttr( "IsNonabelianSimpleGroup", IsSimpleGroup and IsPerfectGroup );
+# ... implies not abelian, not nilpotent (also not solvable, but
+# HasIsSolvableGroup only gets defined later)
 InstallTrueMethod( HasIsAbelian, IsNonabelianSimpleGroup );
+InstallTrueMethod( HasIsNilpotentGroup, IsNonabelianSimpleGroup );
```

Comment:
I have not seen this technique used before. Can you just confirm just
```
InstallTrueMethod( HasIsX, IsY );
```
means that objects `o` for which `IsY(o) = true` becomes known then immediately have the knowledge that `IsX(o) = false`?

Is this because the "default" value of a property is somehow `false`, and that setting the tester for the property to be `true` basically means that this "default" value becomes the accepted truth?

Reply:
No, that's not what happens. If `IsY(o) = true` then it stays true.

But if that happens, you are in an inconsistent state. It typically happens because an overeager user invokes `Set IsNonabelianSimpleGroup(G,true)` on a group that is not actually nonabelian simple. As with other footguns in GAP, we don't attempt to do anything about it.

We *could* in theory at least set a flag that indicate an inconsistency. I guess that's what the `IsImpossible` attribute is meant for (off-side: why is that an attribute, not a property?!). But to use it, one currently needs to add immediate methods, which incurs a performance penalty. Or one makes a generic such method, but then gets less helpful errors (just that an inconsistency occurred, but not how)

Reply:
I don't quite answer @wilfwilson 's question, but here is a demonstration that GAP does get the answer false if it doesn't have any methods:

```
DeclareProperty("IsInteresting",IsGroup);;
DeclareProperty("IsVeryInteresting",IsGroup);
DeclareProperty("IsBoring",IsGroup);

# Every very interesting group is interesting
InstallTrueMethod(IsInteresting, IsVeryInteresting);

# An interesting group is not boring
InstallTrueMethod(HasIsBoring, IsInteresting);
```

test it:

```
gap> g := SL(2,3);;
gap> SetIsVeryInteresting(g,true);
gap> HasIsBoring(g);
true
gap> IsBoring(g);
false
```

Reply:
Ah, I see now what you were getting at, sorry, I misunderstood. So the point is that GAP stores filters as bits in a huge bitmask stored in the type of any object. So if an object is in that filter, that bit is set, otherwise it is not. 

For properties, there are two filters/flags, and hence two bits: one stores whether the value is known or not; the other stores the actual value.

Initially all flag/filter bits are zero. So when we set `HasIsFOO` to true, but do not touch `IsFOO` , we leave the latter at its default value, which is 0 / false.

It would certainly be nice to have a nicer API for this, an `InstallFalseMethod` if you will (see issue #235 where exactly this was requested in 2015 -- but @mohamed-barakat already requested this way earlier on the GAP mailing list, going back to at least 2011)

But in a nutshell, the GAP implication system is too primitive to provide a general `InstallFalseMethod` right now. But what we could do is provide a version that supports some special cases; namely `InstallFalseMethod(IsFOO, cond)` where `IsFOO` is a "simple" property, i.e., declared via `DeclareProperty`, and *not* the result if AND-ing multiple properties. Because for those, we can do `SetIsFOO(obj, false)`. We can't do that for filters like `IsAbelian and IsFinite` in general, because what should that do? It could mean the group is not abelian, or it could mean it is not finite, or both (well, one *could* handle some more cases if one wanted, by translating e.g. `InstallFalseMethod(IsAbelian and IsFinite, req);` to 
```gap
InstallTrueMethod(HasIsAbelian, req and IsFinite);
InstallTrueMethod(HasIsFinite, req and IsAbelian);
```
but getting there (and making sure one does not screw up) requires some non-trivial amount of work; not sure if it's really worth it? (It would be  useful if one also implemented the rest requested in issue #235, namely support for "extended" implications, but that's a much bigger change)

Reply:
Thanks for all that, those first few paragraphs are exactly the info that I was missing.